src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -86,86 +139,138 @@ namespace NovelRT::Graphics
+        {
+            if (buffer->GetLargestFreeRegionSize() >= sizeToStage)
+            {
+                chosenBuffer = buffer;
+                break;
+            }
+        }
 
-        if (stagingBuffer->GetLargestFreeRegionSize() < sizeToStage)
+        if (chosenBuffer == nullptr)
         {
-            _stagingBufferSize = (_stagingBufferSize * 2) + sizeToStage;
-            std::shared_ptr<GraphicsBuffer> newStagingBuffer = CreateStagingBuffer();
-            currentContext->Copy(newStagingBuffer, stagingBuffer);
-            currentContext->EndFrame();
-            _graphicsDevice->Signal(currentContext->GetFence());
-            _graphicsDevice->WaitForIdle();
-            _stagingBuffer.reset(newStagingBuffer);
-            stagingBuffer = _stagingBuffer.getActual();
-            currentContext->BeginFrame();
+            chosenBuffer = CreateStagingBuffer();
         }
-        return stagingBuffer;
+
+        return chosenBuffer;
     }
 
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::CreateStagingBuffer()
     {
-        return _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
+        auto bufferToReturn = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
             Graphics::GraphicsBufferKind::Default, Graphics::GraphicsResourceAccess::Write,
             Graphics::GraphicsResourceAccess::Read, _stagingBufferSize);
+
+        _stagingBuffers[_contextIndex].emplace_back(bufferToReturn);
+
+        return bufferToReturn;
     }
 
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetOrCreateGraphicsBufferForAllocationSize(
-        size_t allocationSize)
+        size_t allocationSize,
+        GraphicsBufferKind bufferKind)
     {
-        for (auto&& buffer : _vertexBuffers)
+        std::vector<std::shared_ptr<GraphicsBuffer>>* buffers = nullptr;
+
+        switch (bufferKind)
+        {
+            case GraphicsBufferKind::Default:
+                buffers = &_defaultBuffers;
+                break;
+            case GraphicsBufferKind::Vertex:
+                buffers = &_vertexBuffers;
+                break;
+            case GraphicsBufferKind::Index:
+                buffers = &_indexBuffers;
+                break;
+            case GraphicsBufferKind::Constant:
+                buffers = &_constantBuffers;
+                break;
+        }
+
+        for (auto&& buffer : *buffers)
         {
             if (buffer->GetLargestFreeRegionSize() >= allocationSize)
             {
                 return buffer;
             }
         }
 
-        size_t minimumBlockSize = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
-        size_t sizeToAllocate = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
+        std::shared_ptr<GraphicsBuffer> newBuffer = nullptr;
 
-        while (sizeToAllocate < allocationSize)
+        if (bufferKind == GraphicsBufferKind::Constant)
```

Comment:
I wonder if you could do something like `newBuffer = bufferKind == constant ? createConstant() : [&]{ /* create correct buffer and return it */ }();`.

Reply:
You want a whole new private method for creating a constant buffer? Or?

Reply:
No, I simplified the code for creating a constant buffer to `createConstant()` since it's a single method call. My bad :sweat_smile:

Reply:
To clarify, `createConstant` was a placeholder for lines 241-243.

Reply:
OK, so I've got it as a ternary, but its now an unreadable mess. You sure you want this?